uaclient/security.py

```diff
@@ -0,0 +1,397 @@
+API_V1_NOTICE_TMPL = "notices/{notice}.json"
+
+
+class SecurityAPIError(util.UrlError):
+    def __init__(self, e, error_response):
+        super().__init__(e, e.code, e.headers, e.url)
+        self.message = error_response.get("message", "")
+
+    def __contains__(self, error_code):
+        return bool(error_code in self.message)
+
+    def __get__(self, error_str, default=None):
+        if error_str in self.message:
+            return self.message
+        return default
+
+    def __str__(self):
+        prefix = super().__str__()
+        details = [self.message]
+        if details:
+            return prefix + ": [" + self.url + "] " + ", ".join(details)
+        return prefix + ": [" + self.url + "]"
+
+
+class UASecurityClient(serviceclient.UAServiceClient):
+
+    cfg_url_base_attr = "security_url"
+    api_error_cls = SecurityAPIError
+
+    def get_cves(
+        self,
+        q: "Optional[str]" = None,
+        priority: "Optional[str]" = None,
+        package: "Optional[str]" = None,
+        limit: "Optional[int]" = None,
+        offset: "Optional[int]" = None,
+        component: "Optional[str]" = None,
+        version: "Optional[str]" = None,
+        status: "Optional[List[str]]" = None,
+    ):
+        """Query to match multiple-CVEs.
+
+        @return: Dict of the JSON response containing the response.
+        """
+        query_params = {
+            "q": q,
+            "priority": priority,
+            "package": package,
+            "limit": limit,
+            "offset": offset,
+            "component": component,
+            "version": version,
+            "status": status,
+        }
+        cves_response, _headers = self.request_url(
+            API_V1_CVES, query_params=query_params
+        )
+        return [CVE(client=self, response=cve_md) for cve_md in cves_response]
+
+    def get_cve(self, cve_id: str):
+        """Query to match single-CVE.
+
+        @return: CVE instance for JSON response from the Security API.
+        """
+        cve_response, _headers = self.request_url(
+            API_V1_CVE_TMPL.format(cve=cve_id)
+        )
+        return CVE(client=self, response=cve_response)
+
+    def get_notices(
+        self,
+        details: "Optional[str]" = None,
+        release: "Optional[str]" = None,
+        limit: "Optional[int]" = None,
+        offset: "Optional[int]" = None,
+        order: "Optional[str]" = None,
+    ):
+        """Query to match multiple-CVEs.
+
+        @return: Dict of the JSON response containing the response.
+        """
+        query_params = {
+            "details": details,
+            "release": release,
+            "limit": limit,
+            "offset": offset,
+            "order": order,
+        }
+        usns_response, _headers = self.request_url(
+            API_V1_NOTICES, query_params=query_params
+        )
+        return [USN(client=self, response=usn_md) for usn_md in usns_response]
+
+    def get_notice(self, notice_id: str):
```

Comment:
I think my explanation was not clear, sorry for that. I  understand that we have different query params for both notices and cves, but the idea would be to mantain the signature of those methods as is and do something like this:

```py
    def get_cves(
        self,
        query: "Optional[str]" = None,
        priority: "Optional[str]" = None,
        package: "Optional[str]" = None,
        limit: "Optional[int]" = None,
        offset: "Optional[int]" = None,
        component: "Optional[str]" = None,
        version: "Optional[str]" = None,
        status: "Optional[List[str]]" = None,
    ) -> "List[CVE]":
        """Query to match multiple-CVEs.
        @return: List of CVE instances based on the the JSON response.
        """
        query_params = {
            "q": query,
            "priority": priority,
            "package": package,
            "limit": limit,
            "offset": offset,
            "component": component,
            "version": version,
            "status": status,
        }
        security_cls = CVE
        request_fmt = API_V1_CVES
        return self.query_multiple_items(query_params, request_fmt, security_cls)

    def get_notices(
        self,
        details: "Optional[str]" = None,
        release: "Optional[str]" = None,
        limit: "Optional[int]" = None,
        offset: "Optional[int]" = None,
        order: "Optional[str]" = None,
    ) -> "List[USN]":
        """Query to match multiple-USNs.
        @return: List of USN instances based on the the JSON response.
        """
        query_params = {
            "details": details,
            "release": release,
            "limit": limit,
            "offset": offset,
            "order": order,
        }
        security_cls = USN
        request_fmt = API_V1_NOTICES
        return self.query_multiple_items(query_params, request_fmt, security_cls)

    def query_multiple_items(self, query_params, request_fmt, security_cls):
        response, _headers = self.request_url(
            request_fmt, query_params=query_params
        )
        return [security_cls(client=self, response=md) for md in response]
 ```
 
 And this same idea also for `get_cve` and `get_notice`. The main idea here is to have a common function that performs the querys

Reply:
Thanks @lucasmoura. I still feel this is a bit of overkill just because we now have a level of indirection on two very simple methods.  I think we talked post-standup about this and it sounded like you were ok with the initial cut (given that we will probably need to use the batch get_cves and get_notices methods to reduce the number of API round trips when multiple CVEs are related to a single USN or multiple USNs related to a single CVE.